framework/src/auxkernels/NodalPatchRecovery.C

```diff
@@ -20,11 +20,11 @@ validParams<NodalPatchRecovery>()
                              orders,
                              "Polynomial order used in least squares fitting of material property "
                              "over the local patch of elements connected to a given node");
-  params.addParamNamesToGroup("patch_polynomial_order", "Advanced");
 
-  // TODO make this class work with relationship manager
-  // params.registerRelationshipManagers("ElementSideNeighborLayers");
-  // params.addPrivateParam<unsigned int short>("element_side_neighbor_layers", 2);
+  params.registerRelationshipManagers("ElementSideNeighborLayers");
+  params.addParam<unsigned short>("element_side_neighbor_layers", 2, "Number of layers to ghost");
```

Comment:
I think we should ghost out N layers of elements, where N is _mesh.dim()

Reply:
No - That makes absolutely no sense. Why?! You always need exactly one level for FE assembly, regardless of dimension. At the point where you want more levels, 2 is a good place to start and can give you basically what you need for patch recovery for any dimension. There's no relation between layers and dimension.

Reply:
Maybe I misunderstood what ElementSideNeighborLayers RM does? I thought, just by looking at its name, it adds relations for off-processor elements that share at least one common side with some on-processor element.

For example, if we have a four-element mesh like this:
1   2
3   4
where element 1 is on processor 1, and elements 2, 3, 4 are on processor 2.
What I thought is that the RM would add algebraic info of element 2 and 3 onto processor 1 if we set element_side_neighbor_layers = 1
but for the lower right corner node of element 1, it actually needs element 4, so I will likely get a segfault when I am on that node.

Reply:
I think  "element_side_neighbor_layers = 2" should work for all dimensional (1D, 2D, and 3D) problems. If you guys can send me a test that shows this does not work, I will fix it.

Reply:
@hugary1995 I stopped by @fdkong's desk with a "test"/demo made out of sticky note blocks to explain about catching the corner node shared by two "diagonal" hex elements, but I don't have an input file

Reply:
OK, talked to @sapitts. I am now understanding what you guys are asking for. I will use a visualization tool to confirm if the ghosted information is enough.

Reply:
> For example, if we have a four-element mesh like this:
1 2
3 4
where element 1 is on processor 1, and elements 2, 3, 4 are on processor 2.
What I thought is that the RM would add algebraic info of element 2 and 3 onto processor 1 if we set element_side_neighbor_layers = 1
but for the lower right corner node of element 1, it actually needs element 4, so I will likely get a segfault when I am on that node.

Actually this is all correct and exactly what the default coupling does in libMesh. You already get one level of geometric and algebraic information. However, if you need something attached at a point, rather than adding more layers of element side neighbor layers (so of which you need to get the diagonal element, but some of what you don't, say elements further away) you'd set up an RM that would just get you exactly what you need, those attached at a specific point. That RM does exist by the way. It's called "ElementPointNeighbors". The point of the system is to be much more explicit about exactly what you need instead of relying on the side effects of something else. You can control how much geometry you need from every point along with how much solution information, down to exactly which variables you need and which ones you don't.

Reply:
Actually Fande, now I understand why @hugary1995 wanted 3 layers for three dimensions. If you need the "off-plane" diagonal element from a hex, you actually need 3 side neighbor layers. The downside is that you ghost a lot more than you need just to pick up that extra element. The worse part is that if your mesh is unstructured, that still may not be enough!

The right solution is to ask specifically for point connected neighbors in the first place.

Reply:
It may be a good idea to add PointNeighbors instead of sideNeighbors.

Reply:
Sorry I wasn't aware that we already had ElementPointNeighbors RM... I think if we make ElementPointNeighbors inherit from AlgebraicRelationshipManager then it should be exactly what we want.

Reply:
Right!